File: medium/3532-path-existence-queries-in-a-graph-i/solution.py
from typing import List


class Solution:
    def pathExistenceQueries(self, n: int, nums: List[int], maxDiff: int, queries: List[List[int]]) -> List[bool]:
        """
        3532. Path Existence Queries in a Graph I
        Time: O(n + q)
        Space: O(n)
        """
        # Component id: increments whenever a consecutive gap exceeds maxDiff.
        comp = [0] * n
        for i in range(1, n):
            comp[i] = comp[i - 1] + (1 if nums[i] - nums[i - 1] > maxDiff else 0)

        return [comp[u] == comp[v] for u, v in queries]

File: medium/3532-path-existence-queries-in-a-graph-i/solution.java
/**
 * 3532. Path Existence Queries in a Graph I
 * Time: O(n + q)
 * Space: O(n)
 */
class Solution {
    public boolean[] pathExistenceQueries(int n, int[] nums, int maxDiff, int[][] queries) {
        // Component id: increments whenever a consecutive gap exceeds maxDiff.
        int[] comp = new int[n];
        for (int i = 1; i < n; i++) {
            comp[i] = comp[i - 1] + (nums[i] - nums[i - 1] > maxDiff ? 1 : 0);
        }

        boolean[] ans = new boolean[queries.length];
        for (int i = 0; i < queries.length; i++) {
            ans[i] = comp[queries[i][0]] == comp[queries[i][1]];
        }
        return ans;
    }
}
